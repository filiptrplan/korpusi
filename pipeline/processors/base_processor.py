class BaseProcessor:
    """
    Base class for all processors. 
    All processors should inherit from this class and implement the process method.
    """
    def __init__(self, song: any, name: str, mapping = None):
        """    
        any song: The song to process. 
        str name: The name of the processor. This is the name of the field that the results will be stored in.
        """
        self.song = song
        self.name = name
        self.mapping = mapping

    def process(self):
        """The main function of the processor. It should spit out the results in dictionary format."""
        raise NotImplementedError

    def get_name(self):
        """Returns the name of the processor."""
        return self.name
    
    def get_mapping(self):
        """Returns the mapping of the processor."""
        if self.mapping is None:
            raise ValueError("Mapping not set")
        return self.mapping