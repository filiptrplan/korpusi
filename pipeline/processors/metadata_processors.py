import csv
import os.path


class CSVMetadataProcessor:
    csv_content = []
    csv_header = []

    def __init__(self, csv_file_path: str):
        """
        @param csv_file_path: This should be the path of the CSV file containing the metadata. The first row should be the
        header specifying different columns. This class will match the `filename` column and then output the other columns
        as metadata. Call the `process` function to get the metadata for a specific file.
        """
        separator = ","
        with open(csv_file_path, "r", encoding="utf-8") as f:
            first_line = f.readline()
            if "," in first_line:
                separator = ","
            elif ";" in first_line:
                separator = ";"
        with open(csv_file_path, "r", encoding="utf-8") as f:
            reader = csv.DictReader(f, delimiter=separator)
            for row in reader:
                self.csv_content.append(row)
            self.csv_header = reader.fieldnames
            if "filename" not in self.csv_header:
                raise ValueError("CSV file must have a column named 'filename'")

    def get_mapping(self):
        properties_dict = {
            "filename": {"enabled": False},
        }

        for column in self.csv_header:
            if column != "filename":
                properties_dict[column] = {
                    "type": "keyword",
                    "fields": {"text": {"type": "text"}},
                }

        return {"properties": properties_dict}

    def process(self, file_path: str):
        """
        Returns the metadata for the file.
        """
        filename = os.path.basename(file_path)
        # Get only the file name without the extension
        filename = os.path.splitext(filename)[0]
        for row in self.csv_content:
            # Get the row filename without the extension
            row_filename = os.path.splitext(row["filename"])[0]
            if row_filename == filename:
                return row
        return {}
