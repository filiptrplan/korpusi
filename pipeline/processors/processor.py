import music21

class Processor:
    """
    Base class for all processors. 
    All processors should inherit from this class and implement the process method.
    """
    def __init__(self, song: music21.stream.Stream, name: str):
        """    
        music21.stream.Stream song: The song to process. This should be output from music21.converter.parse.

        str name: The name of the processor. This is the name of the field that the results will be stored in.
        """
        self.song = song
        self.name = name


    def process(self):
        """The main function of the processor. It should spit out the results in JSON."""
        raise NotImplementedError
    
    
    def get_results(self):
        """Returns the results of the process function in JSON."""
        raise NotImplementedError


    def get_name(self):
        """Returns the name of the processor."""
        return self.name
