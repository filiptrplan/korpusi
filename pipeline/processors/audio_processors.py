import os
import wave

import soundfile

from processors.base_processor import BaseProcessor


class AudioProcessor(BaseProcessor):
    """
    Base class for all audio processors.
    All processors that process audio files should inherit from this class and implement the process method.
    """

    def __init__(self, song: any, name: str = None, mapping=None):
        """
        any song: The path to the song to process.
        str name: The name of the processor. This is the name of the field that the results will be stored in.
        """
        if not isinstance(song, str):
            raise ValueError("Song path must be a string")
        if not os.path.exists(song):
            raise ValueError("Song path does not exist")
        if not song.endswith('.wav'):
            raise ValueError("Song must be a .wav file. Please re-encode any other formats to WAV.")
        if name is None:
            name = self.__class__.__name__
        super().__init__(song, name, mapping)


class AudioFileInfoProcessor(AudioProcessor):
    """Gets the file information of the song like the duration, sample rate, and bit rate."""

    def __init__(self, song: any, name='sample_rate'):
        super().__init__(song, name)
        self.mapping = {
            'properties': {
                'sample_rate': {
                    'type': 'float'
                },
                'duration': {
                    'type': 'float'
                },
                'bit_depth': {
                    'type': 'float'
                }
            }
        }

    def process(self):
        file = soundfile.SoundFile(self.song)
        print(file.subtype)
        return {
            'sample_rate': file.samplerate,
            'duration': file.frames / file.samplerate,
            'bit_depth': file.subtype[0]
        }


class AudioBPMProcessor(AudioProcessor):
    """Gets the BPM of the song."""

    def __init__(self, song: any, name='bpm'):
        super().__init__(song, name)
        self.mapping = {
            'properties': {
                'bpm': {
                    'type': 'float'
                },
                'beat_ticks': {
                    'type': 'float'
                }
            }
        }

