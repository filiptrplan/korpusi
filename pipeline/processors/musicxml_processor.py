import music21
from processors.base_processor import BaseProcessor


class MusicXMLProcessor(BaseProcessor):
    """
    Base class for all musicXML processors.
    All processors that process musicXML files should inherit from this class and implement the process method.
    """

    def __init__(self, song: music21.stream.Stream, name: str = None, mapping=None):
        """
        music21.stream.Stream song: The song to process. This should be output from music21.converter.parse.

        The name of the processor is automatically set to the name of the class. It should be hard-coded in the class to
        something meaningful to better identify the results.
        """
        if name is None:
            name = self.__class__.__name__
        super().__init__(song, name, mapping)

    def process(self):
        """The main function of the processor. It should spit out the results in dictionary format or a single value."""
        raise NotImplementedError

    def get_name(self):
        """Returns the name of the processor."""
        return self.name
