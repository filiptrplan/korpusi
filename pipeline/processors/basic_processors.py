from music21 import stream, tempo, note, interval, metadata, key
from processors.musicxml_processor import MusicXMLProcessor


class KeyProcessor(MusicXMLProcessor):
    """Get the key signature of the song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="key"):
        super().__init__(song, feature_name)
        self.mapping = {
            "properties": {
                "most_certain_key": {
                    "type": "keyword",
                    "fields": {"text": {"type": "text"}},
                },
                "alternate_keys": {
                    "type": "keyword",
                    "fields": {"text": {"type": "text"}},
                },
            }
        }

    def process(self) -> dict[str, list[str] | str]:
        song_key: key.Key = self.song.analyze("key")
        return {
            "most_certain_key": song_key.tonicPitchNameWithCase,
            "alternate_keys": [
                x.tonicPitchNameWithCase for x in song_key.alternateInterpretations[0:4]
            ],
            # get the first 4 alternate keys
        }


class TimeSignatureProcessor(MusicXMLProcessor):
    """Get all time signatures in the song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="time_signature"):
        super().__init__(song, feature_name)
        self.mapping = {"type": "keyword", "fields": {"text": {"type": "text"}}}

    def process(self) -> list[str]:
        time_signatures = set()

        # Iterate through all parts and measures to find time signatures
        for part in self.song.parts:
            measures = part.getElementsByClass(stream.Measure)
            for measure in measures:
                if measure.timeSignature is not None:
                    time_signatures.add(str(measure.timeSignature.ratioString))

        # Convert set to sorted list for consistent ordering
        return sorted(list(time_signatures))


class TempoProcessor(MusicXMLProcessor):
    """Get the tempo of the song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="tempo"):
        super().__init__(song, feature_name)
        self.mapping = {"type": "long"}

    def process(self) -> str:
        tempo_marks = []
        for sub_stream in self.song.recurse(streamsOnly=True, includeSelf=True):
            found = sub_stream.getElementsByClass(tempo.MetronomeMark)
            for x in found:
                tempo_marks.append(x)

        if len(tempo_marks) == 0:
            tempo_str = None
        elif tempo_marks[0].number is not None:
            tempo_str = str(tempo_marks[0].number)
        else:
            tempo_str = None
        return tempo_str


class AmbitusProcessor(MusicXMLProcessor):
    """Gets the 'ambitus' of a song. This is the range of the song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="ambitus"):
        super().__init__(song, feature_name)
        self.mapping = {
            "properties": {
                "min_note": {
                    "type": "long",
                },
                "max_note": {
                    "type": "long",
                },
                "ambitus_semitones": {"type": "long"},
            }
        }

    def process(self):
        notes = self.song.flatten().getElementsByClass(note.Note)
        min_note = note.Note(
            "C8"
        )  # je tole v redu? Ponavadi gre od 0 do 8 po oktavah navzgor
        max_note = note.Note("A0")
        for n in notes:
            max_note = max(n, max_note)
            min_note = min(n, min_note)

        my_interval = interval.Interval(noteStart=min_note, noteEnd=max_note)

        result = {
            "min_note": min_note.pitch.midi,
            "max_note": max_note.pitch.midi,
            "ambitus_semitones": my_interval.semitones,
        }

        return result


class MetadataProcessor(MusicXMLProcessor):
    """Gets the metadata of a song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="metadata"):
        super().__init__(song, feature_name)
        self.mapping = {
            "type": "object",
        }

    def process(self):
        my_metadata = self.song.getElementsByClass(metadata.Metadata)[0]
        my_metadata_tuples = my_metadata.all()
        my_metadata_dict = {}
        for t in my_metadata_tuples:
            my_metadata_dict[t[0]] = t[1]
        my_metadata_dict.pop("filePath")
        my_metadata_dict.pop("fileFormat")
        my_metadata_dict.pop("software")
        return my_metadata_dict


class DurationProcessor(MusicXMLProcessor):
    """Gets the duration of a song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="duration"):
        super().__init__(song, feature_name)
        self.mapping = {
            "properties": {
                "measures": {"type": "long"},
                "beats": {"type": "long"},
            }
        }

    def process(self):
        beats = self.song.duration.quarterLength
        measures = 0
        for x in self.song.flatten():
            if x.measureNumber is not None and x.measureNumber > measures:
                measures = x.measureNumber
        return {"measures": measures, "beats": beats}
