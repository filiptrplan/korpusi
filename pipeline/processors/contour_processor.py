import music21

from processors.musicxml_processor import MusicXMLProcessor

class ContourProcessor(MusicXMLProcessor):
    """Gets the contour of the song. It describes the shape of the melody(when notes ascend or descend)."""

    song: music21.stream.Stream
    def __init__(self, song: music21.stream.Stream, name='contour'):
        super().__init__(song, name)
        self.mapping = {
            'properties': {
                'melodic_contour_semi': {
                    'type': 'long'
                },
                'melodic_contour_string': {
                    'type': 'text'
                },
                'measure_starts': {
                    'type': 'long'
                }
            }
        }

    def process(self):
        pitch_values = []
        measure_numbers = [] # this is done because you can't set a measure number to a note and a note in a chord doesn't have one
        beats = []
        for x in self.song.parts[0].flatten():
            if len(measure_numbers) != 0:
                if measure_numbers[len(measure_numbers) - 1] == x.measureNumber and beats[len(beats) - 1] == x.beat:
                    continue # prevents adding two notes that are played at the same time
            if isinstance(x, music21.note.Note):
                pitch_values.append(x)
                measure_numbers.append(x.measureNumber)
                beats.append(x.beat)
            elif isinstance(x, music21.chord.Chord):
                note = x.sortAscending()[len(x) - 1]
                measure_numbers.append(x.measureNumber)
                pitch_values.append(note)
                beats.append(x.beat)
        melodic_contour = []
        melodic_contour_string = ""
        for i in range(len(pitch_values) - 1):
            pitch_interval = music21.interval.Interval(pitch_values[i], pitch_values[i+1])
            semi = pitch_interval.semitones
            melodic_contour.append(semi)
            
        melodic_contour_string_absolute = ' '.join([x.nameWithOctave for x in pitch_values])
            
        melodic_contour_string = ""
        for x in melodic_contour:
            if x > 0: 
                melodic_contour_string += "U"
            elif x < 0: 
                melodic_contour_string += "D"
            else: 
                melodic_contour_string += "S"
            melodic_contour_string += " " # this is here so we can better search for the contour

        measure_starts = [i for i in range(len(pitch_values)) if measure_numbers[i] != measure_numbers[i-1]]

        return {
            'melodic_contour_semi': melodic_contour,
            'melodic_contour_string_relative': melodic_contour_string,
            'melodic_contour_string_absolute': melodic_contour_string_absolute,
            'measure_starts': measure_starts
        }
    