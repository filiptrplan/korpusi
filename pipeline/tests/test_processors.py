import music21

from processors.basic_processors import (
    DurationProcessor,
    MetadataProcessor,
    AmbitusProcessor,
    TempoProcessor,
    TimeSignatureProcessor,
    KeyProcessor,
)


def song():
    return music21.converter.parse("tests/test.musicxml")


class TestBasicProcessors:
    """Tests processors from basic_processors.py"""

    def test_key_processor(self, snapshot):
        key_processor = KeyProcessor(song())
        assert key_processor.process() == snapshot

    def test_time_signature_processor(self):
        time_signature_processor = TimeSignatureProcessor(song())
        assert time_signature_processor.process() == "3/8"

    def test_tempo_processor(self):
        tempo_processor = TempoProcessor(song())
        assert tempo_processor.process() == "120"

    def test_ambitus_processor(self):
        ambitus_processor = AmbitusProcessor(song())
        assert ambitus_processor.process() == {
            "min_note": 55,
            "max_note": 60,
            "ambitus_semitones": 5,
        }

    def test_metadata_processor(self, snapshot):
        metadata_processor = MetadataProcessor(song())
        assert metadata_processor.process() == snapshot

    def test_duration_processor(self, snapshot):
        duration_processor = DurationProcessor(song())
        assert duration_processor.process() == snapshot
