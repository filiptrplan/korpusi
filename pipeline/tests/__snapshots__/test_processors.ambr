# serializer version: 1
# name: TestBasicProcessors.test_duration_processor
  dict({
    'beats': 49.5,
    'measures': 33,
  })
# ---
# name: TestBasicProcessors.test_key_processor
  dict({
    'alternate_keys': list([
      'd',
      'c',
      'g',
      'G',
    ]),
    'most_certain_key': 'C',
  })
# ---
# name: TestBasicProcessors.test_metadata_processor
  dict({
    'composer': 'Lovrenc Arniƒç',
    'movementName': 'test.musicxml',
    'title': 'Mati',
  })
# ---
